src/popover/internal.tsx

```diff
@@ -57,6 +57,8 @@ function InternalPopover(
   const clickFrameId = useRef<number | null>(null);
 
   const [visible, setVisible] = useState(false);
+  const [wasVisible, setWasVisible] = useState(false);
+  const [dismissedViaDocumentClick, setDismissedViaDocumentClick] = useState(false);
```

Comment:
it is possible to simplify the state by replacing `visible` with some `displayState` union, like:

```
type DisplayState = { type: 'hidden' } | { type: 'visible' } | { type: 'dismissed', withClick: boolean }
```